File: library/src/main/java/org/incendo/interfaces/interfaces/CombinedInterface.java
package org.incendo.interfaces.interfaces;

import net.kyori.adventure.text.Component;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.incendo.interfaces.click.ClickHandler;
import org.incendo.interfaces.pane.CombinedPane;
import org.incendo.interfaces.transform.AppliedTransform;
import org.incendo.interfaces.view.CombinedInterfaceView;
import org.incendo.interfaces.view.InterfaceView;

import java.util.Collection;
import java.util.Map;

public record CombinedInterface(
    int rows,
    Component initialTitle,
    Map<InventoryCloseEvent.Reason, CloseHandler> closeHandlers,
    Collection<AppliedTransform<CombinedPane>> transforms,
    Collection<ClickHandler> clickPreprocessors
) implements Interface<CombinedPane>, TitledInterface {

    @Override
    public int totalRows() {
        return this.rows + 4;
    }

    @Override
    public CombinedPane createPane() {
        return new CombinedPane(this.rows);
    }

    @Override
    public InterfaceView open(final Player player, final InterfaceView parent) {
        CombinedInterfaceView view = new CombinedInterfaceView(player, this, parent);
        view.open();

        return view;
    }
}

File: next/src/main/java/org/incendo/interfaces/next/interfaces/CombinedInterface.kt
package org.incendo.interfaces.next.interfaces

import net.kyori.adventure.text.Component
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.inventory.ItemStack
import org.incendo.interfaces.next.click.ClickHandler
import org.incendo.interfaces.next.pane.CombinedPane
import org.incendo.interfaces.next.transform.AppliedTransform
import org.incendo.interfaces.next.view.CombinedInterfaceView
import org.incendo.interfaces.next.view.InterfaceView

public class CombinedInterface internal constructor(
    override val rows: Int,
    override val initialTitle: Component?,
    override val closeHandlers: MutableMap<InventoryCloseEvent.Reason, CloseHandler>,
    override val transforms: Collection<AppliedTransform<CombinedPane>>,
    override val clickPreprocessors: Collection<ClickHandler>,
    override val itemPostProcessor: ((ItemStack) -> Unit)?
) : Interface<CombinedPane>, TitledInterface {

    override fun totalRows(): Int = rows + 4

    override fun createPane(): CombinedPane = CombinedPane(rows)

    override suspend fun open(player: Player, parent: InterfaceView?): CombinedInterfaceView {
        val view = CombinedInterfaceView(player, this, parent)
        view.open()

        return view
    }
}
